### PR TITLE
Setup general application skeleton

### PR DESCRIPTION
**Pull request by wkozyra95**
Minimal application skeleton

- Add HTTP server that listens for API calls
- Add TCP socket that can be used to connect input and output connections
  - one thread to listen for new connections
  - one thread for each input to listen for incoming data from tcp connection
  - one thread for each output to listen on a channel and write data to tcp connection
- Add Pipeline abstraction that is responsible for processing video (from rtp on input to rtp on output, excluding transport layer)
  - Passing data to the pipeline is done just via a function call (all the processing before the queue will be done synchronously)
  - Passing output data from the pipeline is done via the channel (but when we implement everything I suspect we can remove that)

Not tested in any way,

### CODE REVIEW

File: src/http.rs [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        Request::RegisterOutput(RegisterOutputRequest { port, output_id }) => {
+            state.add_output(port, output_id)?;
+            Ok(())
+        }
+    }
+}
+
+fn handle_request(state: Arc<State>, request: &mut tiny_http::Request) -> Response<impl Read> {
+    let event = match serde_json::from_reader::<_, Request>(request.as_reader()) {
```
Comment (wkozyra95) [inline src/http.rs:45]: obviously better error handling is need here

File: src/pipeline.rs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+                eprintln!("render loop");
+                thread::sleep(Duration::from_millis(1000));
+            }
+        });
+    }
+}
+
+impl PipelineInput {
+    fn process_input(&self, buffer: bytes::Bytes) -> Result<Vec<Frame>> {
+        self.rtp_parser
```
Comment (wkozyra95) [inline src/pipeline.rs:107]: no idea if parsing rtp and decoding will be 2 separate elements or maybe we will go from rtp to raw frames directly. This POC does not address how the timestamp would be passed there.

File: src/state.rs [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+#[allow(dead_code)]
+pub struct PendingConnection {
+    pub port: u16,
+    pub tcp_stream: TcpStream,
+}
+
+pub struct SyncHashMap<K, V>(std::sync::Mutex<HashMap<K, V>>);
+
+impl<K, V> SyncHashMap<K, V>
```
Comment (wkozyra95) [inline src/state.rs:36]: We could use some lib, but here we are dealing with very small maps, so this approach is good enough, and this way we can customize it more for our needs.

File: src/state.rs [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+            pending_connections: SyncHashMap::new(),
+            pipeline,
+        }
+    }
+
+    pub fn add_output(&self, port: u16, output_id: u32) -> Result<()> {
+        let pending_output = self
+            .pending_connections
+            .remove(&port)
+            .ok_or_else(|| anyhow!("no pending connection for port {}", port))?;
```
Comment (WojciechBarczynski) [inline src/state.rs:87]: If we want to return some useful information in the HTTP response, maybe we should use [thiserror](https://docs.rs/thiserror/latest/thiserror/) here.

File: src/pipeline.rs [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+                //
+                // let input_frames = pipeline.queue.next();
+                // let pipeline.render.render(output_frames);
+                // for let (output_id, frames) in input_frames {
+                //     self.on_output_data_received(output_id, frames)
+                // }
+                eprintln!("render loop");
+                thread::sleep(Duration::from_millis(1000));
+            }
+        });
```
Comment (WojciechBarczynski) [inline src/pipeline.rs:101]: In the future, we might want to implement different queuing strategies, that better fit certain use cases. With this render loop, queues have to produce frames in stable intervals, which might not always be the case. I think that this should be a queue responsibility to "push" frames further in the pipeline.

**Review by WojciechBarczynski — CHANGES_REQUESTED** [review]

File: src/decoder.rs [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+pub struct Decoder {}
+
+impl Decoder {
+    pub fn new() -> Self {
+        Decoder {}
+    }
+
+    pub fn decode(&self, _buffer: bytes::Bytes) -> Result<Vec<bytes::Bytes>> {
+        todo!()
+    }
```
[the remark this answers is not in the document]
Comment (wkozyra95) [reply src/decoder.rs:12]: I would leave it as minimal as it is, I expect stuff like this to change when implementing it, so adding stuff like that now just adds work to future refactor.

File: src/state.rs [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    net::TcpStream,
+    sync::{Arc, Mutex},
+};
+
+use crate::{pipeline::Pipeline, tcp_connections};
+
+#[allow(dead_code)]
+pub struct Frame {
+    pub data: bytes::Bytes,
+}
```
[the remark this answers is not in the document]
Comment (WojciechBarczynski) [reply src/state.rs:14]: It's useful to add here `resolution` as well. The `renderer` needs the information about frame resolution to upload them correctly to GPU.

File: src/pipeline.rs (hunk 5 above)
Comment (wkozyra95) [reply src/pipeline.rs:101]: it is in this implementation, I expect that next call to be blocking.

Comment (WojciechBarczynski) [reply src/pipeline.rs:101]: I got it wrong. Now I understand what you meant here :) The `.next()` will block the `queue`

File: src/state.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (wkozyra95) [reply src/state.rs:14]: This is just a placeholder, I expect this type will be defined in the renderer or queu code

File: src/decoder.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (WojciechBarczynski) [reply src/decoder.rs:12]: Ok

**Review by WojciechBarczynski — APPROVED** [review]
LGTM

**Review by jerzywilczek — APPROVED** [review]
:rocket:

**Comment by wkozyra95** [comment]
Closing in favor of https://github.com/membraneframework/video_compositor/pull/17

**Pull request closed without merging**